in_toto/attestations.go

```diff
@@ -131,3 +142,126 @@ type CycloneDXStatement struct {
 	StatementHeader
 	Predicate interface{} `json:"predicate"`
 }
+
+/*
+An Attestor is used to create in-toto Attestation
+framework compliant metadata.
+
+  - GenerateStatement exists to enable usage of alternative
+    signing solutions.
+  - Attest supports using the DSSE for signing
+*/
+type Attestor interface {
+	GenerateStatement() (*v1.Statement, error)
+	Attest() (*Envelope, error)
+}
+
+/*
+LinkAttestor follows the in-toto Attestation framework
+for a Link Predicate.
+*/
+type LinkAttestor struct {
+	MaterialPaths     []string
+	ProductPaths      []string
+	HashAlgorithms    []string
+	GitignorePatterns []string
+	LStripPaths       []string
+	CmdArgs           []string
+	RunDir            string
+	StepName          string
+	LineNormalization bool
+	FollowSymlinkDirs bool
+	Key               Key
+}
+
+func (a LinkAttestor) Attest() (*Envelope, error) {
+	statement, err := a.GenerateStatement()
+	if err != nil {
+		return nil, err
+	}
+
+	return SignStatement(statement, a.Key)
+}
+
+func (a LinkAttestor) GenerateStatement() (*v1.Statement, error) {
+	materials, err := RecordArtifacts(a.MaterialPaths, a.HashAlgorithms, a.GitignorePatterns, a.LStripPaths, a.LineNormalization, a.FollowSymlinkDirs)
+	if err != nil {
+		return nil, err
+	}
+
+	// make sure that we only run RunCommand if cmdArgs is not nil or empty
+	byProducts := map[string]interface{}{}
+	if len(a.CmdArgs) != 0 {
+		byProducts, err = RunCommand(a.CmdArgs, a.RunDir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	products, err := RecordArtifacts(a.ProductPaths, a.HashAlgorithms, a.GitignorePatterns, a.LStripPaths, a.LineNormalization, a.FollowSymlinkDirs)
+	if err != nil {
+		return nil, err
+	}
+
+	var materialResources []*v1.ResourceDescriptor
+	for key, value := range materials {
+		materialResources = append(materialResources, &v1.ResourceDescriptor{
+			Name:   key,
+			Digest: value,
+		})
+	}
+
+	var productResources []*v1.ResourceDescriptor
+	for key, value := range products {
+		productResources = append(productResources, &v1.ResourceDescriptor{
+			Name:   key,
+			Digest: value,
+		})
+	}
+
+	byProductsStruct, err := structpb.NewStruct(byProducts)
+	if err != nil {
+		return nil, err
+	}
+
+	link := link.Link{
+		Name:       a.StepName,
+		Command:    a.CmdArgs,
+		Materials:  materialResources,
+		Byproducts: byProductsStruct,
+	}
+
+	linkJson, err := protojson.Marshal(&link)
+	if err != nil {
+		return nil, err
+	}
+
+	linkStruct := &structpb.Struct{}
+	err = protojson.Unmarshal(linkJson, linkStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	statement := v1.Statement{
+		Type:          StatementInTotoV1,
+		Subject:       productResources,
+		PredicateType: PredicateLink,
+		Predicate:     linkStruct,
+	}
+
+	return &statement, nil
+}
+
+func SignStatement(statement *v1.Statement, key Key) (*Envelope, error) {
```

Comment:
It's possible to make this private, however I thought it may be useful for signing unsigned statements. For example, there are a few tools outputting unsigned statements. A command could be added to pass an unsigned statement to this function to sign it. So, I left it public for future usage. Cosign has a command called "attest" that does something similar.

Let me know what you think.

Reply:
I think you're right, but I lean towards staring off making this private and then making it public after we've nailed down what we want to provide with this so there's API consistency.